simtools/util/workflow_description.py

```diff
@@ -0,0 +1,571 @@
+            if relevant fields are not defined in toplevel metadata
+            dictionary
+
+        """
+
+        _file_format = "ascii.ecsv"
+        try:
+            _file_format = self.workflow_config['PRODUCT']['FORMAT']
+        except KeyError:
+            self._logger.info(
+                "Using default file format for model file: ascii.ecsv")
+
+        if suffix and _file_format == "ascii.ecsv":
+            _file_format = "ecsv"
+
+        return _file_format
+
+    def product_data_directory(self):
+        """
+        Return output directory for data products.
+        Create directory if necessary.
+
+        Output directory is determined following this sorted
+        list:
+        1. PRODUCT:DIRECTORY is set (e.g., through workflow file)
+        2. gammasim-tools output location
+
+        Returns
+        -------
+        str
+            output directory for data products
+
+        Raises
+        ------
+        KeyError
+            if data file name is not defined in workflow configuration
+            or in product metadata dict
+
+        """
+
+        _output_label = self.workflow_config['ACTIVITY']['NAME']
+
+        if self.workflow_config['PRODUCT']['DIRECTORY']:
+            path = Path(self.workflow_config['PRODUCT']['DIRECTORY'])
+            path.mkdir(parents=True, exist_ok=True)
+            _output_dir = path.absolute()
+        else:
+            _output_dir = cfg.get("outputLocation")
+
+        _output_dir = io.getApplicationOutputDirectory(
+            _output_dir, _output_label)
+
+        self._logger.info("Outputdirectory {}".format(_output_dir))
+        return _output_dir
+
+    def _from_args(self, key, default_return=None):
+        """
+        Return argparser argument.
+        No errors raised if argument does not exist
+
+        """
+
+        try:
+            return self.args.__dict__[key]
+        except KeyError:
+            pass
+
+        return default_return
+
+    def _fill_toplevel_meta_from_args(self):
+        """
+        Fill metadata available through command line into top-level template
+
+        Parameters
+        ----------
+        args: argparse.Namespace
+            command line parameters
+
+        Raises
+        ------
+        KeyError
+            if metadata description cannot be filled
+
+        """
+
+        try:
+            _association = {}
+            _association['SITE'] = self.args.site
+            _split_telescope_name = self.args.telescope.split("-")
+            _association['CLASS'] = _split_telescope_name[0]
+            _association['TYPE'] = _split_telescope_name[1]
+            _association['SUBTYPE'] = _split_telescope_name[2]
+            self.toplevel_meta['CTA']['PRODUCT']['ASSOCIATION'][0] = _association
+        except KeyError:
+            self._logger.error("Error reading user input meta data from args")
+            raise
+        except AttributeError:
+            pass
+
+    def _fill_toplevel_meta_from_file(self):
```

Comment:
I must admit I did not check that all of the possible entries in the top level metadata schema are included in this function. If you would like me to do that as well, please let me know.

Also, not sure why `toplevel` is one word in the name of the method (appears in other places as well).

Reply:
This is clearly a weak point, that a change in args has some impact here. I think we covered everything necessary here.

More 'import': so you would prefer to replace toplevel by top_level everywhere? I really don't care, fine with me.

Reply:
Would you hate me if I ask to change also `userinput` to `user_input` everywhere? Otherwise that's the last comment, everything is fine, we can merge.

Reply:
Done - this was only for one method.